[CHORE] 린트 설정, rest docs 설정

park-se-jun opened the pull request:
## PR Description 🗒️
1. naver lint 설정
2. rest docs 의 산출물을 openapi3 스펙으로 받고, swagger UI 플러그인으로 통해 swagger UI를 생성해서 확인할 수 있도록 함
  - 절차
    1. restdocs test 작성
    3. restdocs-api-spec이 openapi3 스펙으로 결과물을 산출
    4. openapi3 스펙을 바탕으로 swagger UI 플러그인이 swagger 관련 정적 리소스 생성
    5. bootJar 시 정적 리소스를 jar 파일에 포함
    6. localhost/docs/index.html을 통해 swagger화면을 볼 수 있음




## 추가/변경 사항 및 설명 📝
1. 린트 설정 관련
- naver 컨벤션을 맞출 수 있도록 하는 xml 추가
- intelij 포매터 naver-intellij-formatter.xml 추가 [적용방법 D.2](https://naver.github.io/hackday-conventions-java/#_intellij)
- checkstyle 플러그인 설정 방법 [적용방법 D.2.6](https://naver.github.io/hackday-conventions-java/#_intellij)
- checkstyle 하는  workflow 작성

2. rest docs 설정 관련
   - rest docs 설정
   - rest docs를 openapi3 스펙으로 변경해주는 플러그인 [restdocs-api-spec](https://github.com/ePages-de/restdocs-api-spec) 을 이용해 openapi3 스펙이 나오도록 함
   - openapi3 스펙을 swagger UI로 만들어주는 플러그인 [swagger-generator](https://github.com/int128/gradle-swagger-generator-plugin)사용 swagger 정적 파일 생성
   - 위의 프로세스를 bootJar시 자동으로 resources에 포함되도록 build.gradle 수정
   - test용 healthcheck api 작성
   - 참고자료
     - [자료1](https://blog.jdriven.com/2021/10/generate-swagger-ui-from-spring-rest-docs/)
     - [자료2](https://jwkim96.tistory.com/274)
     - [restdocs-api-spec](https://github.com/ePages-de/restdocs-api-spec) 
     - [swagger-generator-plugin](https://github.com/int128/gradle-swagger-generator-plugin)

## To Reviewers 🙏 
- 각자의 로컬에 꼭 formatter를 설정 해 주시기 바랍니다.
- rest docs를 이용해 컨트롤러 테스트를 진행할때 아래의 스니펫처럼 작성하여야 합니다. 참고해 작성해주세요
```java
mockMvc.perform(get("/carts/{id}", cartId)
  .accept(HAL_JSON))
  .andExpect(status().isOk())
  .andDo(document("cart-get",
    resource(ResourceSnippetParameters.builder()
      .description("Get a cart by id")
      .pathParameters(
        parameterWithName("id").description("the cart id"))
      .responseFields(
        fieldWithPath("total").description("Total amount of the cart."),
        fieldWithPath("products").description("The product line item of the cart."),
        subsectionWithPath("products[]._links.product").description("Link to the product."),
        fieldWithPath("products[].quantity").description("The quantity of the line item."),
        subsectionWithPath("products[].product").description("The product the line item relates to."),
        subsectionWithPath("_links").description("Links section."))
      .links(
        linkWithRel("self").ignored(),
        linkWithRel("order").description("Link to order the cart."))
    .build())));
```
## Issue close ✂️
close #6 
close #5

park-se-jun committed:
[CHORE] 린트를 하기 위한 xml,.editorconfig, build.gradle 추가
- intelij 용 formatter xml 추가
- checkstyle을 위한 xml 추가
- 필요한 플러그인들 추가

park-se-jun committed:
[ADD] checkstyle을 검사하는 workflow 작성

park-se-jun committed:
[CHORE] 발생할 수도 있는 crlf, lf 문제를 위한 파일

park-se-jun committed:
[CHORE] sejun 으로부터 merge

park-se-jun committed:
[CHORE] workflow 오류 수정

park-se-jun committed:
[CHORE] : workflow에 java 설정

park-se-jun committed:
[CHORE] sejun으로부터 merge

park-se-jun committed:
[CHORE] pr시에는 checkstyle이 안돌아가도록 변

park-se-jun committed:
Merge pull request #11 from prgrms-web-devcourse/sejun_#6

[CHORE] 린트 설정

twotwobread reviewed (commented):
소현님이 deploy.yml 만들 때 runner 설정이랑 checkout, permission, jdk 등 ci 환경을 만드는데 똑같은 환경을 다른 YAML 파일에서 만들면 테스트하는데 시간이 더 걸리는 부분은 없나요?? deploy.yml에 다 넣는게 나을까요 이렇게 분리 시키는게 나을까요??

twotwobread commented on .github/workflows/checkstyle.yml:
```diff
@@ -0,0 +1,21 @@
+  check-style:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
소현님은 checkout@v3을 사용하고 세준님은 checkout@v2를 사용했는데 특별한 이유가 있나요??

thguss replied:
@twotwobread 저는 가장 최신 버전을 사용했습니다 ㅋㅋ

park-se-jun replied:
그저 복붙....

thguss commented on build.gradle:
```diff
@@ -1,97 +1,112 @@
-	id 'io.spring.dependency-management' version '1.0.15.RELEASE'
-	id 'org.asciidoctor.jvm.convert' version '3.2.0'
-	id 'jacoco'
+    id 'java'
```
이거 탭 수가 바뀐 것 같은데 확인 부탁드려요! 아마 저와 수영님이 lint 설정을 안해서 발생한 일 같은데,, 우선 이전 버전으로 돌려주시고 린트 관련 이슈를 따로 파서 고치는 게 좋을 것 같아 보입니다!

thguss replied:
일단 build.gradle 파일 모두 이전 버전으로 복구 부탁드려요 ㅠㅠ

park-se-jun replied:
마지막에 merge 하기로 합의 완료

thguss reviewed (dismissed):
수고하셨습니다~!

twotwobread reviewed (dismissed):
이해완료했슴다!! 수고했습니다!!

park-se-jun committed:
[CHORE] rest docs + swagger 설정 테스트

- rest docs + swager 관련 플러그인 및 의존성 추가

park-se-jun committed:
[FEAT] 테스트용 health check api 작성

park-se-jun committed:
[TEST] health check api controller 테스트 추

park-se-jun committed:
[CHORE] check style의 예외로 test의 네이밍 추가
- test 코드들의 메소드 네이밍을 한글로 하기로 정했기 때문에, naver 컨벤션과 맞지 않는 문제가 발생해 test code에 한해서 적용시키지 않도록 설

park-se-jun committed:
[CHORE] check style 관련 오류 수정
- checkstyle suppression 설정 오류 수정
- gitattributes에 bat 파일의 end를 crlf로 되게정끔 설

park-se-jun committed:
[CHORE] check style 의 NewlineAtEndOfFile 위반 수

park-se-jun committed:
[CHORE] rest docs와 swagger를 합치기 위한 task 설정 정

- 사용하지 않는 postman task 삭제
- swagger의 결과물의 이름을 ApiDocument로 변경
- GenerateSwaggerUI가 Openapi3를 먼저 실행시키도록 설정
- generateSwaggerUIApiDocument를 먼저 진행하도록 하는 copySwaggerUI 태스크 생성
- 해당 테스트에서 swagger ui의 산출물을 build/resorce/main/static/docs로 이동
- 생성한 copySwaggerUI 테스크가 bootJar이전에 실행되도록 의존성 설정
- check에 걸려 있던 editorconfigCheck 주석화(bat file의 eol 관련 이슈 발생)

park-se-jun committed:
[TEST] security 비활성화, swagger 사용을 하기 위한 코드 변경

- 테스트시 security auto config 때문에 302가 나오는 문제로 인해 일시적 security 비활성화
- openapi 스펙에 맞추기 휘해 document 메소드 내부의 parameter를 resource()를 사용하도록 변경(추가 설치한 라이브러리에서 제공)

park-se-jun committed:
Merge pull request #24 from prgrms-web-devcourse/sejun_#5

## PR Description 🗒️
- rest docs 의 산출물을 openapi3 스펙으로 받고, swagger UI 플러그인으로 통해 swagger UI를 생성해서 확인할 수 있도록 함
절차
1. restdocs test 작성
2. restdocs-api-spec이 openapi3 스펙으로 결과물을 산출
3. openapi3 스펙을 바탕으로 swagger UI 플러그인이 swagger 관련 정적 리소스 생성
4. bootJar 시 정적 리소스를 jar 파일에 포함
5. localhost/docs/index.html을 통해 swagger화면을 볼 수 있음
## 추가/변경 사항 및 설명 📝
- rest docs 설정
- rest docs를 openapi3 스펙으로 변경해주는 플러그인 [restdocs-api-spec](https://github.com/ePages-de/restdocs-api-spec) 을 이용해 openapi3 스펙이 나오도록 함
- openapi3 스펙을 swagger UI로 만들어주는 플러그인 [swagger-generator](https://github.com/int128/gradle-swagger-generator-plugin)사용 swagger 정적 파일 생성
- 위의 프로세스를 bootJar시 자동으로 resources에 포함되도록 build.gradle 수정
- test용 healthcheck api 작성

참고자료
- [자료1](https://blog.jdriven.com/2021/10/generate-swagger-ui-from-spring-rest-docs/)
- [자료2](https://jwkim96.tistory.com/274)
- [restdocs-api-spec](https://github.com/ePages-de/restdocs-api-spec) 
- [swagger-generator-plugin](https://github.com/int128/gradle-swagger-generator-plugin)
## To Reviewers 🙏 // 중점적 봐줬으면 하는 부분.

## Issue close ✂️

thguss commented on src/test/java/com/example/demo/controller/HealthControllerTest.java:
```diff
@@ -0,0 +1,57 @@
+	private MockMvc mockMvc;
+
+	@Test
+	public void 성공_헬스체크_요청을했을때_특정문자열반환() throws Exception {
```
테스트 컨벤션을 정확히 합의하고 넘어가지 않았어서 간단히 얘기해보면 좋을 것 같아요 :)

thguss reviewed (dismissed):
수고하셨습니다~!

twotwobread commented on src/test/java/com/example/demo/controller/HealthControllerTest.java:
```diff
@@ -0,0 +1,57 @@
+@ExtendWith(RestDocumentationExtension.class)
+@AutoConfigureRestDocs
+@WebMvcTest(controllers = HealthController.class, excludeAutoConfiguration = {SecurityAutoConfiguration.class})
+class HealthControllerTest {
```
이 테스트 스위트는 무엇을 위한 테스트 스위트죠?? RestDocs가 잘적용되는지 확인해보는 테스트 스위트라면 이 부분이 repo에 코드로 올라올 필요는 없다고 생각됩니다.

park-se-jun replied:
rest docs 와 swagger 설정이 제대로 되었는지 확인 + 서버가 살아있는지를 확인하는 health check 테스트 코드입니다. 서버가 살아있는지 확인할 필요도 있을 것 같기 때문에, health check를 지울게 아니면 뺄 필요는 없을 것 같습니다.

twotwobread commented on config/checkstyle/naver-checkstyle-suppressions.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0"?>
```
이 파일이 스타일 강제하는 파일일까요??

thguss committed:
Merge branch 'develop' into sejun

thguss committed:
Merge branch 'develop' into sejun

thguss reviewed (approved):
허가!

park-se-jun committed:
[CHORE] checkstyle.yml을 deploy.yml로 합치

park-se-jun committed:
Merge pull request #26 from prgrms-web-devcourse/sejun_#25

[CHORE] checkstyle.yml을 deploy.yml로 합치기

twotwobread reviewed (approved):
허가요!
